File: peacedesign/src/main/java/com/peacedesign/android/widget/compound/PeaceCompoundButton.java
/*
 * Copyright (c) Faisal Khan (https://github.com/faisalcodes)
 * Created on 2/4/2022.
 * All rights reserved.
 */

package com.peacedesign.android.widget.compound;

import static android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE;
import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.TextUtils;
import android.text.style.TextAppearanceSpan;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Checkable;
import android.widget.CompoundButton;
import android.widget.LinearLayout;

import androidx.annotation.CallSuper;
import androidx.annotation.IntDef;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.annotation.StyleRes;
import androidx.appcompat.widget.AppCompatTextView;

import com.peacedesign.R;
import com.peacedesign.android.utils.Dimen;
import com.peacedesign.android.utils.ViewUtils;
import com.peacedesign.android.utils.interfaceUtils.CompoundButtonGroup;
import com.peacedesign.android.utils.span.TypefaceSpan2;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public abstract class PeaceCompoundButton extends LinearLayout implements Checkable, CompoundButton.OnCheckedChangeListener {
    public static final int COMPOUND_TEXT_LEFT = 0;
    public static final int COMPOUND_TEXT_TOP = 1;
    public static final int COMPOUND_TEXT_RIGHT = 2;
    public static final int COMPOUND_TEXT_BOTTOM = 3;

    public static final int COMPOUND_TEXT_GRAVITY_START = 0;
    public static final int COMPOUND_TEXT_GRAVITY_END = 1;
    public static final int COMPOUND_TEXT_GRAVITY_CENTER = 2;
    public static final int COMPOUND_TEXT_GRAVITY_LEFT = 3;
    public static final int COMPOUND_TEXT_GRAVITY_RIGHT = 4;

    protected final boolean mInitialChecked;
    @TextGravity
    private int mTextGravity;
    private CompoundButtonGroup mParent;
    private CharSequence mText;
    private CharSequence mSubText;
    @CompoundDirection
    private int mCompoundDirection;
    private int mSpaceBetween;
    private int mTextAppearanceResId;
    private int mSubTextAppearanceResId;

    private BeforeCompoundCheckChangeListener mBeforeCheckListener;
    private OnCompoundCheckChangedListener mCheckChangeListener;

    private AppCompatTextView mTxtView;

    public PeaceCompoundButton(Context context) {
        this(context, null);
    }

    public PeaceCompoundButton(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public PeaceCompoundButton(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public PeaceCompoundButton(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.PeaceCompoundButton, defStyleAttr, defStyleRes);

        mText = a.getText(R.styleable.PeaceCompoundButton_android_text);
        mSubText = a.getText(R.styleable.PeaceCompoundButton_peaceComp_subText);
        mInitialChecked = a.getBoolean(R.styleable.PeaceCompoundButton_android_checked, false);
        mCompoundDirection = a.getInt(R.styleable.PeaceCompoundButton_peaceComp_direction, COMPOUND_TEXT_RIGHT);
        mSpaceBetween = a.getDimensionPixelSize(R.styleable.PeaceCompoundButton_peaceComp_spaceBetween, Dimen.dp2px(context, 10));
        mTextGravity = a.getInt(R.styleable.PeaceCompoundButton_peaceComp_textGravity, COMPOUND_TEXT_GRAVITY_START);
        mTextAppearanceResId = a.getResourceId(R.styleable.PeaceCompoundButton_android_textAppearance,
                R.style.PeaceRadioTextAppearance);
        mSubTextAppearanceResId = a.getResourceId(R.styleable.PeaceCompoundButton_peaceComp_subTextAppearance,
                R.style.PeaceRadioSubTextAppearance);

        a.recycle();
        init();
    }

    private void init() {
        initThis();
        makeComponents();
    }

    @CallSuper
    protected void initThis() {
        super.setOrientation(HORIZONTAL);
        super.setOnClickListener(v -> {
            if (mParent != null) mParent.check(getId());
            else toggle();
        });
    }

    @CallSuper
    protected void makeComponents() {
        makeTexts();
        addCompoundButton();
    }

    private void makeTexts() {
        mTxtView = new AppCompatTextView(getContext());
        LayoutParams paramsText = new LayoutParams(WRAP_CONTENT, WRAP_CONTENT);
        paramsText.weight = 1;
        mTxtView.setLayoutParams(paramsText);
        mTxtView.setGravity(resolveTextGravity(mTextGravity));

        setTextAppearancesInternal(mTextAppearanceResId, mSubTextAppearanceResId);
    }

    protected void addCompoundButton() {
        View compoundButton = getCompoundButton();

        ViewUtils.removeView(mTxtView);
        ViewUtils.removeView(compoundButton);

        switch (mCompoundDirection) {
            case COMPOUND_TEXT_LEFT:
            case COMPOUND_TEXT_TOP:
                addView(mTxtView, 0);
                if (compoundButton != null) {
                    addView(compoundButton, 1);
                }
                break;
            case COMPOUND_TEXT_RIGHT:
            case COMPOUND_TEXT_BOTTOM:
            default:
                if (compoundButton != null) {
                    addView(compoundButton, 0);
                }

                addView(mTxtView, compoundButton != null ? 1 : 0);
                break;
        }

        if (mCompoundDirection == COMPOUND_TEXT_TOP || mCompoundDirection == COMPOUND_TEXT_BOTTOM) {
            super.setOrientation(VERTICAL);
        } else {
            super.setOrientation(HORIZONTAL);
        }

        setSpaceBetweenInternal(mSpaceBetween);
    }

    protected abstract CompoundButton getCompoundButton();

    private int resolveTextGravity(@TextGravity int textGravity) {
        switch (textGravity) {
            case COMPOUND_TEXT_GRAVITY_CENTER:
                return Gravity.CENTER;
            case COMPOUND_TEXT_GRAVITY_END:
                return Gravity.END;
            case COMPOUND_TEXT_GRAVITY_LEFT:
                return Gravity.LEFT;
            case COMPOUND_TEXT_GRAVITY_RIGHT:
                return Gravity.RIGHT;
            case COMPOUND_TEXT_GRAVITY_START:
            default:
                return Gravity.START;
        }
    }

    private void setTextInternal(CharSequence text, CharSequence subtext) {
        SpannableStringBuilder ssb = new SpannableStringBuilder();

        if (!TextUtils.isEmpty(text)) {
            SpannableString titleSS = new SpannableString(text);
            setTextAppearanceSpan(titleSS, mTextAppearanceResId, false);
            ssb.append(titleSS);
        }

        if (!TextUtils.isEmpty(subtext)) {
            if (!TextUtils.isEmpty(text)) ssb.append("\n");

            SpannableString subtitleSS = new SpannableString(subtext);
            setTextAppearanceSpan(subtitleSS, mSubTextAppearanceResId, true);
            ssb.append(subtitleSS);
        }

        mTxtView.setText(ssb);
        mTxtView.setVisibility(ssb.length() > 0 ? VISIBLE : GONE);
    }

    private void setTextAppearanceSpan(SpannableString ss, int styleId, boolean isSubText) {
        int txtSizeDef = Dimen.sp2px(getContext(), isSubText ? 14 : 16);
        TypedArray ta = getContext().obtainStyledAttributes(styleId, androidx.appcompat.R.styleable.TextAppearance);


        String family = "sans-serif";
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            Typeface font = ta.getFont(androidx.appcompat.R.styleable.TextAppearance_android_fontFamily);
            if (font != null) {
                ss.setSpan(new TypefaceSpan2(font), 0, ss.length(), SPAN_EXCLUSIVE_EXCLUSIVE);
            } else {
                family = resolveFontFamily(ta.getString(androidx.appcompat.R.styleable.TextAppearance_android_fontFamily));
            }
        } else {
            family = resolveFontFamily(ta.getString(androidx.appcompat.R.styleable.TextAppearance_android_fontFamily));
        }


        int style = ta.getInt(androidx.appcompat.R.styleable.TextAppearance_android_textStyle,
                isSubText ? Typeface.NORMAL : Typeface.BOLD);
        int size = ta.getDimensionPixelSize(androidx.appcompat.R.styleable.TextAppearance_android_textSize, txtSizeDef);
        ColorStateList color = ta.getColorStateList(androidx.appcompat.R.styleable.TextAppearance_android_textColor);
        TextAppearanceSpan txtApSpan = new TextAppearanceSpan(family, style, size, color, null);
        ss.setSpan(txtApSpan, 0, ss.length(), SPAN_EXCLUSIVE_EXCLUSIVE);

        ta.recycle();
    }

    private String resolveFontFamily(String s) {

        if ("sans-serif".equalsIgnoreCase(s) ||
                "sans-serif-light".equalsIgnoreCase(s) ||
                "sans-serif-condensed".equalsIgnoreCase(s) ||
                "sans-serif-black".equalsIgnoreCase(s) ||
                "sans-serif-thin".equalsIgnoreCase(s) ||
                "sans-serif-medium".equalsIgnoreCase(s)) {
            return s;
        }
        return "sans-serif";
    }

    private void setSpaceBetweenInternal(int spaceBetween) {
        LinearLayout.LayoutParams p = (LayoutParams) mTxtView.getLayoutParams();
        p.setMargins(0, 0, 0, 0);
        p.setMarginStart(0);
        p.setMarginEnd(0);

        switch (mCompoundDirection) {
            case COMPOUND_TEXT_LEFT:
                p.setMarginEnd(spaceBetween);
                break;
            case COMPOUND_TEXT_TOP:
                p.bottomMargin = spaceBetween;
                break;
            case COMPOUND_TEXT_RIGHT:
                p.setMarginStart(spaceBetween);
                break;
            case COMPOUND_TEXT_BOTTOM:
                p.topMargin = spaceBetween;
                break;
        }
        mTxtView.setLayoutParams(p);
    }

    private void setTextAppearancesInternal(int textAppearanceResId, int subTextAppearanceResId) {
        mTextAppearanceResId = textAppearanceResId;
        mSubTextAppearanceResId = subTextAppearanceResId;

        setTextInternal(mText, mSubText);
    }

    public void setTextAppearance(@StyleRes int styleResId) {
        mTextAppearanceResId = styleResId;
        setTextAppearances(styleResId, mSubTextAppearanceResId);
    }

    public void setTexts(@StringRes int textResId, @StringRes int subTextResId) {
        setTexts(getContext().getText(textResId), getContext().getText(subTextResId));
    }

    public void setTexts(CharSequence text, CharSequence subText) {
        mText = text;
        mSubText = subText;

        setTextInternal(text, subText);
    }

    public CharSequence getText() {
        return mText;
    }

    public void setText(@StringRes int resId) {
        setText(getContext().getText(resId));
    }

    public void setText(CharSequence text) {
        mText = text;

        setTextInternal(text, mSubText);
    }

    public CharSequence getSubText() {
        return mSubText;
    }

    public void setSubText(@StringRes int resId) {
        setSubText(getContext().getText(resId));
    }

    public void setSubText(CharSequence subText) {
        mSubText = subText;

        setTextInternal(mText, subText);
    }

    public void setCompoundDirection(@CompoundDirection int compoundDirection) {
        mCompoundDirection = compoundDirection;

        addCompoundButton();
    }

    public void setSpaceBetween(int spaceBetween) {
        mSpaceBetween = spaceBetween;
        setSpaceBetweenInternal(spaceBetween);
    }

    public void setTextAppearances(@StyleRes int textAppearanceResId, @StyleRes int subTextAppearanceResId) {
        setTextAppearancesInternal(textAppearanceResId, subTextAppearanceResId);
    }

    public void setSubTextAppearance(@StyleRes int styleResId) {
        mSubTextAppearanceResId = styleResId;
        setTextAppearances(mTextAppearanceResId, styleResId);
    }

    public void setForceTextGravity(@TextGravity int gravity) {
        mTextGravity = gravity;

        if (mTxtView != null) {
            mTxtView.setGravity(resolveTextGravity(gravity));
        }
    }

    public void setGroup(@NonNull CompoundButtonGroup radioGroupPro) {
        mParent = radioGroupPro;
    }

    public void setOnCheckedChangedListener(OnCompoundCheckChangedListener listener) {
        mCheckChangeListener = listener;
    }

    public void setOnBeforeCheckChangedListener(BeforeCompoundCheckChangeListener listener) {
        mBeforeCheckListener = listener;
    }

    @Override
    public boolean isChecked() {
        CompoundButton btn = getCompoundButton();
        if (btn == null) {
            return false;
        }
        return btn.isChecked();
    }

    @Override
    public void setChecked(boolean checked) {
        CompoundButton btn = getCompoundButton();
        if (btn == null) {
            return;
        }

        if (mBeforeCheckListener == null || mBeforeCheckListener.beforeCheck(this, checked)) {
            btn.setChecked(checked);
        }
    }

    @Override
    public void toggle() {
        CompoundButton btn = getCompoundButton();
        if (btn == null) {
            return;
        }
        if (mBeforeCheckListener == null || mBeforeCheckListener.beforeCheck(this, !isChecked())) {
            btn.toggle();
        }
    }

    @CallSuper
    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        Drawable background = getBackground();
        if (background != null) {
            int[] state = {isChecked ? android.R.attr.state_checked : -android.R.attr.state_checked};
            background.setState(state);
        }

        if (mCheckChangeListener != null) {
            mCheckChangeListener.onCheckChanged(this, isChecked);
        }
    }

    @Override
    public void setOrientation(int orientation) {
        // overridden
    }

    @Override
    public boolean onInterceptTouchEvent(@NonNull MotionEvent ev) {
        return true;
    }

    @IntDef({COMPOUND_TEXT_LEFT, COMPOUND_TEXT_TOP, COMPOUND_TEXT_RIGHT, COMPOUND_TEXT_BOTTOM})
    @Retention(RetentionPolicy.SOURCE)
    public @interface CompoundDirection {}

    @IntDef({COMPOUND_TEXT_GRAVITY_START, COMPOUND_TEXT_GRAVITY_END, COMPOUND_TEXT_GRAVITY_CENTER, COMPOUND_TEXT_GRAVITY_LEFT, COMPOUND_TEXT_GRAVITY_RIGHT})
    @Retention(RetentionPolicy.SOURCE)
    public @interface TextGravity {}

    public interface OnCompoundCheckChangedListener {
        void onCheckChanged(PeaceCompoundButton button, boolean isChecked);
    }

    public interface BeforeCompoundCheckChangeListener {
        boolean beforeCheck(PeaceCompoundButton button, boolean newState);
    }
}

File: app/src/main/java/com/quranapp/android/widgets/compound/PeaceCompoundButton.kt
package com.quranapp.android.widgets.compound

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Typeface
import android.os.Build
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.TextUtils
import android.text.style.TextAppearanceSpan
import android.util.AttributeSet
import android.view.Gravity
import android.view.MotionEvent
import android.view.ViewGroup
import android.widget.Checkable
import android.widget.CompoundButton
import android.widget.LinearLayout
import androidx.annotation.CallSuper
import androidx.annotation.IntDef
import androidx.annotation.StringRes
import androidx.annotation.StyleRes
import androidx.appcompat.widget.AppCompatTextView
import com.peacedesign.android.utils.span.TypefaceSpan2
import com.quranapp.android.R
import com.quranapp.android.utils.extensions.dp2px
import com.quranapp.android.utils.extensions.removeView
import com.quranapp.android.utils.extensions.sp2px

abstract class PeaceCompoundButton @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr), Checkable, CompoundButton.OnCheckedChangeListener {
    companion object {
        const val COMPOUND_TEXT_LEFT = 0
        const val COMPOUND_TEXT_TOP = 1
        const val COMPOUND_TEXT_RIGHT = 2
        const val COMPOUND_TEXT_BOTTOM = 3

        const val COMPOUND_TEXT_GRAVITY_START = 0
        const val COMPOUND_TEXT_GRAVITY_END = 1
        const val COMPOUND_TEXT_GRAVITY_CENTER = 2
        const val COMPOUND_TEXT_GRAVITY_LEFT = 3
        const val COMPOUND_TEXT_GRAVITY_RIGHT = 4
    }

    protected var initialChecked = false

    @TextGravity
    private var textGravity = 0
    private var parent: PeaceCompoundButtonGroup? = null
    private var text: CharSequence? = null
    private var subText: CharSequence? = null

    private var compoundDirection = 0
    private var spaceBetween = 0
    private var textAppearanceResId = 0
    private var subTextAppearanceResId = 0

    var beforeCheckChangeListener: ((PeaceCompoundButton, Boolean) -> Boolean)? = null
    var onCheckChangedListener: ((PeaceCompoundButton, Boolean) -> Unit)? = null

    private var txtView: AppCompatTextView? = null

    init {
        val a = context.obtainStyledAttributes(
            attrs,
            R.styleable.PeaceCompoundButton,
            defStyleAttr,
            0
        )

        text = a.getText(R.styleable.PeaceCompoundButton_android_text)
        subText = a.getText(R.styleable.PeaceCompoundButton_peaceComp_subText)
        initialChecked = a.getBoolean(R.styleable.PeaceCompoundButton_android_checked, false)
        compoundDirection = a.getInt(
            R.styleable.PeaceCompoundButton_peaceComp_direction,
            COMPOUND_TEXT_RIGHT
        )
        spaceBetween = a.getDimensionPixelSize(
            R.styleable.PeaceCompoundButton_peaceComp_spaceBetween,
            context.dp2px(10f)
        )
        textGravity = a.getInt(
            R.styleable.PeaceCompoundButton_peaceComp_textGravity,
            COMPOUND_TEXT_GRAVITY_START
        )
        textAppearanceResId = a.getResourceId(
            R.styleable.PeaceCompoundButton_android_textAppearance,
            R.style.PeaceRadioTextAppearance
        )
        subTextAppearanceResId = a.getResourceId(
            R.styleable.PeaceCompoundButton_peaceComp_subTextAppearance,
            R.style.PeaceRadioSubTextAppearance
        )

        a.recycle()
        init()
    }

    private fun init() {
        initThis()
        makeComponents()
    }

    @CallSuper
    protected open fun initThis() {
        super.setOrientation(HORIZONTAL)
        super.setOnClickListener { if (parent != null) parent!!.check(id) else toggle() }
    }

    @CallSuper
    protected open fun makeComponents() {
        makeTexts()
        addCompoundButton()
    }

    private fun makeTexts() {
        txtView = AppCompatTextView(context).apply {
            layoutParams = LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply {
                weight = 1f
            }
            gravity = resolveTextGravity(textGravity)
        }
        setTextAppearancesInternal(textAppearanceResId, subTextAppearanceResId)
    }

    protected open fun addCompoundButton() {
        val compoundButton = getCompoundButton()
        txtView!!.removeView()
        compoundButton?.removeView()

        when (compoundDirection) {
            COMPOUND_TEXT_LEFT, COMPOUND_TEXT_TOP -> {
                addView(txtView, 0)
                if (compoundButton != null) {
                    addView(compoundButton, 1)
                }
            }

            COMPOUND_TEXT_RIGHT, COMPOUND_TEXT_BOTTOM -> {
                if (compoundButton != null) {
                    addView(compoundButton, 0)
                }
                addView(txtView, if (compoundButton != null) 1 else 0)
            }

            else -> {
                if (compoundButton != null) {
                    addView(compoundButton, 0)
                }
                addView(txtView, if (compoundButton != null) 1 else 0)
            }
        }
        if (compoundDirection == COMPOUND_TEXT_TOP || compoundDirection == COMPOUND_TEXT_BOTTOM) {
            super.setOrientation(VERTICAL)
        } else {
            super.setOrientation(HORIZONTAL)
        }
        setSpaceBetweenInternal(spaceBetween)
    }

    abstract fun getCompoundButton(): CompoundButton?

    @SuppressLint("RtlHardcoded")
    private fun resolveTextGravity(@TextGravity textGravity: Int): Int {
        return when (textGravity) {
            COMPOUND_TEXT_GRAVITY_CENTER -> Gravity.CENTER
            COMPOUND_TEXT_GRAVITY_END -> Gravity.END
            COMPOUND_TEXT_GRAVITY_LEFT -> Gravity.LEFT
            COMPOUND_TEXT_GRAVITY_RIGHT -> Gravity.RIGHT
            COMPOUND_TEXT_GRAVITY_START -> Gravity.START
            else -> Gravity.START
        }
    }

    private fun setTextInternal(text: CharSequence?, subtext: CharSequence?) {
        val ssb = SpannableStringBuilder()

        if (!text.isNullOrEmpty()) {
            val titleSS = SpannableString(text)
            setTextAppearanceSpan(titleSS, textAppearanceResId, false)
            ssb.append(titleSS)
        }
        if (!subtext.isNullOrEmpty()) {
            if (!TextUtils.isEmpty(text)) ssb.append("\n")
            val subtitleSS = SpannableString(subtext)
            setTextAppearanceSpan(subtitleSS, subTextAppearanceResId, true)
            ssb.append(subtitleSS)
        }

        txtView!!.text = ssb
        txtView!!.visibility = if (ssb.isNotEmpty()) VISIBLE else GONE
    }

    @SuppressLint("CustomViewStyleable")
    private fun setTextAppearanceSpan(ss: SpannableString, styleId: Int, isSubText: Boolean) {
        val txtSizeDef = context.sp2px(if (isSubText) 14F else 16.toFloat())
        val ta = context.obtainStyledAttributes(styleId, R.styleable.TextAppearance)

        var family = "sans-serif"

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val font = ta.getFont(R.styleable.TextAppearance_android_fontFamily)
            if (font != null) {
                ss.setSpan(TypefaceSpan2(font), 0, ss.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            } else {
                family = resolveFontFamily(
                    ta.getString(R.styleable.TextAppearance_android_fontFamily)
                )
            }
        } else {
            family = resolveFontFamily(ta.getString(R.styleable.TextAppearance_android_fontFamily))
        }

        val style = ta.getInt(
            R.styleable.TextAppearance_android_textStyle,
            if (isSubText) Typeface.NORMAL else Typeface.BOLD
        )

        val size = ta.getDimensionPixelSize(R.styleable.TextAppearance_android_textSize, txtSizeDef)
        val color = ta.getColorStateList(R.styleable.TextAppearance_android_textColor)
        val txtApSpan = TextAppearanceSpan(family, style, size, color, null)

        ss.setSpan(txtApSpan, 0, ss.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)

        ta.recycle()
    }

    private fun resolveFontFamily(s: String?): String {
        return if ("sans-serif".equals(s, ignoreCase = true) ||
            "sans-serif-light".equals(s, ignoreCase = true) ||
            "sans-serif-condensed".equals(s, ignoreCase = true) ||
            "sans-serif-black".equals(s, ignoreCase = true) ||
            "sans-serif-thin".equals(s, ignoreCase = true) ||
            "sans-serif-medium".equals(s, ignoreCase = true)
        ) {
            s!!
        } else {
            "sans-serif"
        }
    }

    private fun setSpaceBetweenInternal(spaceBetween: Int) {
        (txtView!!.layoutParams as LayoutParams).apply {
            setMargins(0, 0, 0, 0)
            marginStart = 0
            marginEnd = 0

            when (compoundDirection) {
                COMPOUND_TEXT_LEFT -> marginEnd = spaceBetween
                COMPOUND_TEXT_TOP -> bottomMargin = spaceBetween
                COMPOUND_TEXT_RIGHT -> marginStart = spaceBetween
                COMPOUND_TEXT_BOTTOM -> topMargin = spaceBetween
            }
        }
        txtView!!.requestLayout()
    }

    private fun setTextAppearancesInternal(textAppearanceResId: Int, subTextAppearanceResId: Int) {
        this.textAppearanceResId = textAppearanceResId
        this.subTextAppearanceResId = subTextAppearanceResId
        setTextInternal(text, subText)
    }

    fun setTextAppearance(@StyleRes styleResId: Int) {
        textAppearanceResId = styleResId
        setTextAppearances(styleResId, subTextAppearanceResId)
    }

    fun setTexts(@StringRes textResId: Int, @StringRes subTextResId: Int) {
        setTexts(context.getText(textResId), context.getText(subTextResId))
    }

    fun setTexts(text: CharSequence?, subText: CharSequence?) {
        this.text = text
        this.subText = subText

        setTextInternal(text, subText)
    }

    fun getText(): CharSequence? {
        return text
    }

    fun setText(@StringRes resId: Int) {
        setText(context.getText(resId))
    }

    fun setText(text: CharSequence) {
        this.text = text
        setTextInternal(text, subText)
    }

    fun getSubText(): CharSequence? {
        return subText
    }

    fun setSubText(@StringRes resId: Int) {
        setSubText(context.getText(resId))
    }

    fun setSubText(subText: CharSequence) {
        this.subText = subText
        setTextInternal(text, subText)
    }

    fun setCompoundDirection(@CompoundDirection compoundDirection: Int) {
        this.compoundDirection = compoundDirection
        addCompoundButton()
    }

    fun setSpaceBetween(spaceBetween: Int) {
        this.spaceBetween = spaceBetween
        setSpaceBetweenInternal(spaceBetween)
    }

    fun setTextAppearances(
        @StyleRes textAppearanceResId: Int,
        @StyleRes subTextAppearanceResId: Int
    ) {
        setTextAppearancesInternal(textAppearanceResId, subTextAppearanceResId)
    }

    fun setSubTextAppearance(@StyleRes styleResId: Int) {
        subTextAppearanceResId = styleResId
        setTextAppearances(textAppearanceResId, styleResId)
    }

    fun setForceTextGravity(@TextGravity gravity: Int) {
        textGravity = gravity
        txtView?.gravity = resolveTextGravity(gravity)
    }

    override fun setTextAlignment(alignment: Int) {
        txtView?.textAlignment = alignment
    }

    fun setGroup(radioGroupPro: PeaceCompoundButtonGroup) {
        parent = radioGroupPro
    }

    override fun isChecked(): Boolean {
        val btn = getCompoundButton() ?: return false
        return btn.isChecked
    }

    override fun setChecked(checked: Boolean) {
        val btn = getCompoundButton() ?: return
        if (beforeCheckChangeListener == null || beforeCheckChangeListener!!.invoke(this, checked)) {
            btn.isChecked = checked
        }
    }

    override fun toggle() {
        val btn = getCompoundButton() ?: return
        if (beforeCheckChangeListener == null || beforeCheckChangeListener!!.invoke(
                this,
                !isChecked
            )
        ) {
            btn.toggle()
        }
    }

    @CallSuper
    override fun onCheckedChanged(buttonView: CompoundButton?, isChecked: Boolean) {
        background?.state = intArrayOf(
            if (isChecked) android.R.attr.state_checked else -android.R.attr.state_checked
        )

        onCheckChangedListener?.invoke(this, isChecked)
    }

    override fun setOrientation(orientation: Int) {
        // overridden
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        return true
    }

    @IntDef(COMPOUND_TEXT_LEFT, COMPOUND_TEXT_TOP, COMPOUND_TEXT_RIGHT, COMPOUND_TEXT_BOTTOM)
    @Retention(AnnotationRetention.SOURCE)
    annotation class CompoundDirection

    @IntDef(
        COMPOUND_TEXT_GRAVITY_START,
        COMPOUND_TEXT_GRAVITY_END,
        COMPOUND_TEXT_GRAVITY_CENTER,
        COMPOUND_TEXT_GRAVITY_LEFT,
        COMPOUND_TEXT_GRAVITY_RIGHT
    )
    @Retention(AnnotationRetention.SOURCE)
    annotation class TextGravity
}
